central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,624 @@
+	reportSnapshotStore reportSnapshotDS.DataStore, notifierDatastore notifierDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor) Scheduler {
+	cronScheduler := cron.New()
+	cronScheduler.Start()
+	ourSchema, err := graphql.ParseSchema(resolvers.Schema(), resolvers.New())
+	if err != nil {
+		panic(err)
+	}
+	return newSchedulerImpl(reportConfigDatastore, reportMetadataStore, reportSnapshotStore, notifierDatastore,
+		deploymentDatastore, watchedImageDatastore, collectionDatastore, collectionQueryRes, notificationProcessor, cronScheduler, ourSchema)
+}
+
+func newSchedulerImpl(reportConfigDatastore reportConfigDS.DataStore, reportMetadataStore reportMetadataDS.DataStore,
+	reportSnapshotStore reportSnapshotDS.DataStore, notifierDatastore notifierDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor, cronScheduler *cron.Cron, schema *graphql.Schema) *scheduler {
+	s := &scheduler{
+		reportConfigToEntryIDs:  make(map[string]cron.EntryID),
+		reportConfigDatastore:   reportConfigDatastore,
+		reportMetadataStore:     reportMetadataStore,
+		reportSnapshotStore:     reportSnapshotStore,
+		notifierDatastore:       notifierDatastore,
+		deploymentDatastore:     deploymentDatastore,
+		watchedImageDatastore:   watchedImageDatastore,
+		collectionDatastore:     collectionDatastore,
+		collectionQueryResolver: collectionQueryRes,
+		notificationProcessor:   notificationProcessor,
+		reportsQueue:            make([]*ReportRequest, 0),
+		sendNewReports:          make(chan bool, 200),
+		runningReportConfigs:    set.NewStringSet(),
+		Schema:                  schema,
+
+		stopper:         concurrency.NewStopper(),
+		cron:            cronScheduler,
+		concurrencySema: semaphore.NewWeighted(int64(env.ReportExecutionMaxConcurrency.IntegerSetting())),
+	}
+	return s
+}
+
+/* Concurrency and scheduling functions */
+
+func (s *scheduler) Start() {
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.WorkflowAdministration)))
+
+	s.queuePendingReports(ctx)
+	s.queueScheduledReports(ctx)
+	go s.runReports()
+	go s.runReports()
+}
+
+func (s *scheduler) Stop() {
+	s.stopper.Client().Stop()
+	err := s.stopper.Client().Stopped().Wait()
+	if err != nil {
+		log.Errorf("Error stopping vulnerability report scheduler : %v", err)
+	}
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
```

Comment:
I don't think you want to use a boolean channel. Even a value of `false` would set this channel and execute the case.

Reply:
Yeah, I wanted to use something like a Signal here, but I think signals are a one time use thing and do not allow buffering multiple signals. So I went with a boolean channel.

Reply:
The scheduler only needs to know whether or not to pick another report. I don't see a point of buffering here as it is error-prone.